File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExpandingArchives.vb
Imports System
Imports System.IO
Imports System.IO.Compression
Imports System.Linq
Imports System.Text

Namespace ClassLibrary1
    Public Class Class1
        Public Sub ExtractArchive(ByVal archive As ZipArchive)
            For Each entry As ZipArchiveEntry In archive.Entries
                entry.ExtractToFile("") ' Noncompliant
'               ^^^^^^^^^^^^^^^^^^^^^^^
            Next

            For i As Integer = 0 To archive.Entries.Count - 1
                archive.Entries(i).ExtractToFile("") ' Noncompliant
            Next

            archive.Entries.ToList().ForEach(Sub(e) e.ExtractToFile("")) ' Noncompliant
'                                                   ^^^^^^^^^^^^^^^^^^^

            ZipFileExtensions.ExtractToDirectory(archive, "") ' Noncompliant
            archive.ExtractToDirectory("") ' Noncompliant
        End Sub

        Public Sub ExtractEntry(ByVal entry As ZipArchiveEntry)
            Dim fullName As String
            Dim stream As Stream

            entry.ExtractToFile("") ' Noncompliant
            entry.ExtractToFile("", True) ' Noncompliant

            ZipFileExtensions.ExtractToFile(entry, "") ' Noncompliant
            ZipFileExtensions.ExtractToFile(entry, "", True) ' Noncompliant

            ZipFile.ExtractToDirectory("", "") ' Noncompliant
            ZipFile.ExtractToDirectory("", "", Encoding.Default) ' Noncompliant

            stream = entry.Open() ' Compliant, method is not tracked

            entry.Delete() ' Compliant, method is not tracked

            fullName = entry.FullName ' Compliant, properties are not tracked

            ExtractToFile(entry) ' Compliant, method is not tracked
        End Sub

        Public Sub ExtractToFile(ByVal entry As ZipArchiveEntry)
        End Sub
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ExpandingArchives.cs
using System;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Text;

namespace ClassLibrary1
{
    public class Class1
    {
        public void ExtractArchive(ZipArchive archive)
        {
            foreach (var entry in archive.Entries)
            {
                entry.ExtractToFile(""); // Noncompliant
//              ^^^^^^^^^^^^^^^^^^^^^^^
            }

            for (int i = 0; i < archive.Entries.Count; i++)
            {
                archive.Entries[i].ExtractToFile(""); // Noncompliant
            }

            archive.Entries.ToList()
                .ForEach(e => e.ExtractToFile("")); // Noncompliant
//                            ^^^^^^^^^^^^^^^^^^^

            ZipFileExtensions.ExtractToDirectory(archive, ""); // Noncompliant
            archive.ExtractToDirectory(""); // Noncompliant
        }

        public void ExtractEntry(ZipArchiveEntry entry)
        {
            string fullName;
            Stream stream;

            entry.ExtractToFile(""); // Noncompliant
            entry?.ExtractToFile(""); // Noncompliant
            entry.ExtractToFile("", true); // Noncompliant

            ZipFileExtensions.ExtractToFile(entry, ""); // Noncompliant
            ZipFileExtensions.ExtractToFile(entry, "", true); // Noncompliant

            ZipFile.ExtractToDirectory("", ""); // Noncompliant
            ZipFile.ExtractToDirectory("", "", Encoding.Default); // Noncompliant

            stream = entry.Open(); // Compliant, method is not tracked

            entry.Delete(); // Compliant, method is not tracked

            fullName = entry.FullName; // Compliant, properties are not tracked

            ExtractToFile(entry); // Compliant, method is not tracked

            Invoke(ZipFileExtensions.ExtractToFile); // Compliant, not an invocation, but could be considered as FN
        }

        public void ExtractToFile(ZipArchiveEntry entry) { }

        public void Invoke(Action<ZipArchiveEntry, string> action) { }
    }
}
